src/non_expressive_names.rs

```diff
@@ -0,0 +1,279 @@
+                                                                     &interned_name[split..]));
+                                   }
+                               });
+            return;
+        }
+        self.0.names.push((interned_name, span, count));
+    }
+}
+
+impl<'a, 'b> SimilarNamesLocalVisitor<'a, 'b> {
+    /// ensure scoping rules work
+    fn apply<F: for<'c> Fn(&'c mut Self)>(&mut self, f: F) {
+        let n = self.names.len();
+        let single_char_count = self.single_char_names.len();
+        f(self);
+        self.names.truncate(n);
+        self.single_char_names.truncate(single_char_count);
+    }
+}
+
+impl<'v, 'a, 'b> visit::Visitor<'v> for SimilarNamesLocalVisitor<'a, 'b> {
+    fn visit_local(&mut self, local: &'v Local) {
+        if let Some(ref init) = local.init {
+            self.apply(|this| visit::walk_expr(this, &**init));
+        }
+        // add the pattern after the expression because the bindings aren't available yet in the init expression
+        SimilarNamesNameVisitor(self).visit_pat(&*local.pat);
+    }
+    fn visit_block(&mut self, blk: &'v Block) {
+        self.apply(|this| visit::walk_block(this, blk));
+    }
+    fn visit_arm(&mut self, arm: &'v Arm) {
+        self.apply(|this| {
+            // just go through the first pattern, as either all patterns bind the same bindings or rustc would have errored much earlier
+            SimilarNamesNameVisitor(this).visit_pat(&arm.pats[0]);
+            this.apply(|this| visit::walk_expr(this, &arm.body));
+        });
+    }
+    fn visit_item(&mut self, _: &'v Item) {
+        // do nothing
+    }
+}
+
+impl EarlyLintPass for NonExpressiveNames {
+    fn check_fn(&mut self, cx: &EarlyContext, _: FnKind, decl: &FnDecl, blk: &Block, _: Span, _: NodeId) {
+        let mut visitor = SimilarNamesLocalVisitor {
+            names: Vec::new(),
+            cx: cx,
+            lint: &self,
+            single_char_names: Vec::new(),
+        };
+        // initialize with function arguments
+        for arg in &decl.inputs {
+            visit::walk_pat(&mut SimilarNamesNameVisitor(&mut visitor), &arg.pat);
+        }
+        // walk all other bindings
+        visit::walk_block(&mut visitor, blk);
+    }
+}
+
+/// precondition: a.len() < b.len()
+fn levenstein_not_1(a_name: &str, b_name: &str) -> bool {
+    debug_assert!(a_name.chars().count() < b_name.chars().count());
+    let mut a_chars = a_name.chars();
+    let mut b_chars = b_name.chars();
+    while let (Some(a), Some(b)) = (a_chars.next(), b_chars.next()) {
+        if a == b {
+            continue;
+        }
+        if let Some(b2) = b_chars.next() {
+            // check if there's just one character inserted
+            if a == b2 && a_chars.eq(b_chars) {
+                return false;
+            } else {
+                // two charaters don't match
+                return true;
+            }
```

Comment:
it's a loop -> no `needless_return`. But yea, no clue why the bool isn't triggering. maybe because there's a return on there ^^